### PR TITLE
[GPU] Add condition check for dynamic shape and onednn_impl in concat_in_place_optimization::match()

### PR DESCRIPTION
**Pull request by wilson-seok**
### Details:
 - Add condition check for dynamic shape and onednn_impl in concat_in_place_optimization::match()
 
### Tickets:
 - 112919

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp [hunk 1]
```diff
@@ -86,7 +87,7 @@ struct typed_primitive_impl_ocl : public typed_primitive_impl<PType> {
 
     template<typename ImplType>
     static std::unique_ptr<primitive_impl> create(const typed_program_node<PType>& arg, const kernel_impl_params& impl_param) {
-        if (impl_param.can_be_optimized()) {
+        if (impl_param.can_be_optimized() && !(impl_param.typed_desc<concatenation>() && impl_param.is_dynamic())) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:90]: Could you please add note here? 
// concat buffer fusing for dynamic shape is adaptively applied at runtime. So we need to build dynamic impl at build time.

File: src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp [hunk 2]
```diff
@@ -258,6 +259,157 @@ TEST(prepare_buffer_fusing, in_place_concat_dynamic) {
...
+    cldnn::mem_lock<FLOAT16> input2_ptr(input_memory2, get_test_stream());
+
+    const auto& concat_node_n = net.get_primitive("concat")->get_node();
+    auto concat_mem = net.get_primitive("concat")->output_memory_ptr();
+    auto reorder1_mem = net.get_primitive("reorder1")->output_memory_ptr();
+    auto reorder2_mem = net.get_primitive("reorder2")->output_memory_ptr();
+
+    ASSERT_NE(concat_mem.get(), reorder1_mem.get());
+    ASSERT_NE(concat_mem.get(), reorder2_mem.get());
+    ASSERT_TRUE(concat_node_n.can_be_optimized());
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp:406]: Could you add one more check for concat_inst (not node) with can_be_optimized = false

File: src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp [hunk 3]
```diff
@@ -258,6 +259,157 @@ TEST(prepare_buffer_fusing, in_place_concat_dynamic) {
...
+    cldnn::mem_lock<FLOAT16> input2_ptr(input_memory2, get_test_stream());
+
+    const auto& concat_node_n = net.get_primitive("concat")->get_node();
+    auto concat_mem = net.get_primitive("concat")->output_memory_ptr();
+    auto reorder1_mem = net.get_primitive("reorder1")->output_memory_ptr();
+    auto reorder2_mem = net.get_primitive("reorder2")->output_memory_ptr();
+
+    ASSERT_EQ(concat_mem.get(), reorder1_mem.get());
+    ASSERT_EQ(concat_mem.get(), reorder2_mem.get());
+    ASSERT_TRUE(concat_node_n.can_be_optimized());
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp:330]: Could you check concat_inst (not node) is also can_be_optimized = true

Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp:332]: Done. Thanks!

File: src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp (hunk 2 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/unit/passes/prepare_buffer_fusing_test.cpp:410]: Done. Thanks!

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp (hunk 1 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:91]: Done. Thanks!

**Review by yeonbok — APPROVED** [review]

**Pull request merged**